packages/grafana-data/src/transformations/transformers/calculateField.ts

```diff
@@ -180,6 +216,138 @@ export const calculateFieldTransformer: DataTransformerInfo<CalculateFieldTransf
   },
 };
 
+function getWindowCreator(options: WindowOptions, allFrames: DataFrame[]): ValuesCreator {
+  if (options.windowSize! < 1) {
+    throw new Error('Add field from calculation transformation - Window size must be larger than 0');
+  }
+
+  let matcher = getFieldMatcher({
+    id: FieldMatcherID.numeric,
+  });
+
+  if (options.field) {
+    matcher = getFieldMatcher({
+      id: FieldMatcherID.byNames,
+      options: {
+        names: [options.field],
+      },
+    });
+  }
+
+  const info = fieldReducers.get(options.reducer);
+
+  if (!info) {
+    throw new Error(`Unknown reducer: ${options.reducer}`);
+  }
+
+  const reducer = info.reduce ?? doStandardCalcs;
+  const ignoreNulls = options.nullValueMode === NullValueMode.Ignore;
+  const nullAsZero = options.nullValueMode === NullValueMode.AsZero;
+
+  return (frame: DataFrame) => {
+    // Find the columns that should be examined
+    let selectedField: Field | null = null;
+    for (const field of frame.fields) {
+      if (matcher(field, frame, allFrames)) {
+        selectedField = field;
+        break;
+      }
+    }
+
+    if (!selectedField) {
+      return;
+    }
+
+    // Prepare a "fake" field for the row
+    const fakeField: Field = {
+      name: 'temp',
+      values: [],
+      type: FieldType.number,
+      config: {},
+    };
+    const vals: number[] = [];
+
+    if (options.type === WindowType.Centered) {
+      for (let i = 0; i < frame.length; i++) {
+        const boundary = options.windowSize! / 2;
+        const start = Math.ceil(Math.max(0, i - boundary));
+        const end = Math.ceil(Math.min(i + boundary, frame.length));
+
+        fakeField.values = selectedField.values.slice(start, end);
+
+        vals.push(reducer(fakeField, ignoreNulls, nullAsZero)[options.reducer]);
+      }
+    } else {
+      for (let i = 0; i < frame.length; i++) {
+        fakeField.values.push(selectedField.values[i]);
+        if (fakeField.values.length > options.windowSize!) {
+          fakeField.values.shift();
+        }
+
+        vals.push(reducer(fakeField, ignoreNulls, nullAsZero)[options.reducer]);
+      }
```

Comment:
i was gonna say this looks really expensive/wasteful, since for 2000 values (my screen width with random walk) it would have to do 2000 pushes and shifts in addition to running the reducer 2000 times. you could do this much more efficiently by just maintaining a running average that subtracts the value at the trailing index and adds the value at the leading index of the window (for mean). but this does not play nice with existing reducers :(

the code here takes ~1ms to run, which is kind of crazy. JS runtimes are crazy.

i've written the rolling average loop i'm describing for the bottom demo of [1], which does 3600 points in 0.25ms with a window size of 300 (open console for timings).

[1] https://leeoniya.github.io/uPlot/demos/data-smoothing.html

Reply:
you may want to also consider offering a window size as % of total data length rather than a fixed window size.

Reply:
Yes that's what we also figured. We rewrote it to something similar to what you suggested I think. Without using the reducers.